### PR TITLE
Date and slug from filename for leaf bundles

### PR DESCRIPTION
**Pull request by Japanuspus**
The definition of the `BaseFilename` field in the `FrontMatterDescriptor` structure provided to the frontmatter handlers has been changed, so that in the case of a leaf bundle `index.`-file, the name provided is that of the containing bundle folder.
Since the use of the `FrontMatterDescriptor` is localized to the frontmatter handler, this should not have external effects.

To compute the new value of `BaseFilename`, I have added a `ContentBaseName` method to the `File` interface. The `FileInfo` implementation of this interface has been modified to compute the value as part of `lazyinit`. 

This is my first ever `go`-code, so feel free to rewrite as you wish, or let me know if I should implement in another way.

### CODE REVIEW

**Comment by Japanuspus** [comment]
Due to bad github-foo, I am missing a link to the issue: #4558

**Comment by bep** [comment]
@Japanuspus just a note that I have not forgotten this PR, I will get to it in one of the next few days ...

**Comment by moorereason** [comment]
@Japanuspus 
If you're willing and able, you can update your commit message to [follow our guidelines](https://github.com/gohugoio/hugo/blob/master/CONTRIBUTING.md#git-commit-message-guidelines).  We :heart: clean commit messages.  😃

**Comment by bep** [comment]
Other than the before-mentioned commit message, I think this looks fine. I don't love the name of the `ContentBaseName` method, but I cannot come up with anything better. That interface is (from historical reasons) a little too big and slightly confusing, but that is out of scope for this.

@moorereason what do you think?

File: source/fileInfo.go [hunk 1]
```diff
@@ -177,11 +189,15 @@ func (fi *FileInfo) init() {
 		if (!fi.isLeafBundle && len(parts) == 1) || len(parts) > 1 {
 			section = parts[0]
 		}
-
 		fi.section = section
 
-		fi.uniqueID = helpers.MD5String(filepath.ToSlash(fi.relPath))
+		var contentBaseName = fi.translationBaseName
```
Comment (moorereason) [inline source/fileInfo.go]: Simplify this section to avoid writing to `contentBaseName` twice on leaves.

```go
if fi.isLeafBundle && len(parts) > 0 {
    fi.contentBaseName = parts[len(parts)-1]
} else {
    fi.contentBaseName = fi.translationBaseName
}
```

File: source/fileInfo.go [hunk 2]
```diff
@@ -177,11 +189,15 @@ func (fi *FileInfo) init() {
...
+		if fi.isLeafBundle && len(parts) > 0 {
+			contentBaseName = parts[len(parts)-1]
+		}
+		fi.contentBaseName = contentBaseName
 
+		fi.uniqueID = helpers.MD5String(filepath.ToSlash(fi.relPath))
```
Comment (moorereason) [inline source/fileInfo.go]: Unrelated change?

File: source/fileInfo_test.go [hunk 3]
```diff
@@ -94,4 +94,14 @@ func TestFileInfoLanguage(t *testing.T) {
...
+
+	// test contentBaseName implementation
+	fi1 := s.NewFileInfo("", "2018-10-01-contentbasename.md", false, nil)
+	fi2 := s.NewFileInfo("", "2018-10-01-contentbasename.en.md", false, nil)
+	fi3 := s.NewFileInfo("", "2018-10-01-contentbasename"+helpers.FilePathSeparator+"index.en.md",
+		true, nil)
+
+	assert.Equal("2018-10-01-contentbasename", fi1.ContentBaseName())
+	assert.Equal("2018-10-01-contentbasename", fi2.ContentBaseName())
+	assert.Equal("2018-10-01-contentbasename", fi3.ContentBaseName())
```
Comment (moorereason) [inline source/fileInfo_test.go]: Need a test that ContentBaseName equals TranslationBaseName when not in a bundle.

File: source/fileInfo_test.go [hunk 4]
```diff
@@ -94,4 +94,14 @@ func TestFileInfoLanguage(t *testing.T) {
 
 	assert.Equal("sv", fiSv.Lang())
 	assert.Equal("en", fiEn.Lang())
+
+	// test contentBaseName implementation
+	fi1 := s.NewFileInfo("", "2018-10-01-contentbasename.md", false, nil)
+	fi2 := s.NewFileInfo("", "2018-10-01-contentbasename.en.md", false, nil)
+	fi3 := s.NewFileInfo("", "2018-10-01-contentbasename"+helpers.FilePathSeparator+"index.en.md",
```
Comment (moorereason) [inline source/fileInfo_test.go]: Use [`filepath.Join`](https://golang.org/pkg/path/filepath/#Join).

File: source/fileInfo.go (hunk 1 above)
Comment (Japanuspus) [reply source/fileInfo.go]: My initial implementation actually used direct assignment, but then I spotted this local variable pattern in the lines just above (implementing `section`), and thought it might be a golang idiom...

I'll change it back to direct assignment which I also find easier to read.

File: source/fileInfo.go (hunk 2 above)
Comment (Japanuspus) [reply source/fileInfo.go]: Seems to be just a bad diff -- this line was right below my insert. I haven't been able to find out why git is marking this as delete/add.

**Comment by Japanuspus** [comment]
Thanks, @bep and @moorereason. 
I don't quite like the `contentBaseName` either, but `leafName` would obviously not make sense to use in the non-leaf case, so I gave up finding a better alternative.

With respect to the implementation, another approach would be to put the logic in `page.go`, but that seemed much harder to test.

Have just pushed a new commit to this branch without squashing, hope that is ok. New commit message should be usable for PR -- but should I amend the original commit?

**Comment by vassudanagunta** [comment]
The logical name is LogicalName, but that's taken.

**Comment by bep** [comment]
@moorereason is this good to go now?
